autoload/test/rust/cargotest.vim

```diff
@@ -61,28 +61,42 @@ function! s:nearest_test(position) abort
 
   " Else
   " Search forward for the first declared method
-  let name = test#base#nearest_test_in_lines(
+  let name_f = test#base#nearest_test_in_lines(
     \ a:position['file'],
     \ name['test_line'],
     \ a:position['line'],
     \ g:test#rust#cargotest#patterns
   \ )
 
-  return join(name['test'])
+  if len(name['namespace']) > 0
+    return join([name['namespace'][0], name_f['test'][0]], '::')
+  else
+    return name_f['test'][0]
+  endif
 endfunction
 
 function! s:test_namespace(filename) abort
   let l:path = fnamemodify(a:filename, ':r')
   " On a normal cargo project, the first item is 'src'
-  let l:modules = split(l:path, '/')[1:]
+  let l:modules = split(l:path, '/')
 
   " 'src/lib.rs' and 'src/some/mod.rs' does not end
   " with actual module names
   if l:modules[-1] =~# '\v^(lib|mod)$'
     let l:modules = l:modules[:-2]
   endif
 
+
   " Build up tests module namespace
-  let l:modules = l:modules + ['tests']
-  return join(l:modules, '::') . '::'
+  if l:modules[0] == 'tests' && len(l:modules) == 2
+    return ''
+  else
+    let l:modules = l:modules[1:]
+    if len(l:modules) > 0
+      return join(l:modules, '::') . '::'
+    else
+      return ''
+    endif
+  endif
+  " return join(l:modules, '::') . '::'
```

Comment:
Please remove if no longer needed

Reply:
Woops, left it there accidentally. Fixed